overlord/snapstate/snapstate.go

```diff
@@ -550,7 +551,7 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 	}
 
 	// This task is necessary only for UC20+ and hybrid
-	if deviceCtx.HasModeenv() && snapsup.Type == snap.TypeKernel {
+	if snapsup.Type == snap.TypeKernel && needsKernelSetup(deviceCtx) {
```

Comment:
if deviceCtx was potentially the one associated with a task or the remodel code the question below would be addressed I think or at least could be addressed